### PR TITLE
feat(backup): Support comparators

### PR DESCRIPTION
**Pull request by azaslavsky**
This change introduces the idea of comparators, or special handlers for comparing JSON fields on exported models that cannot be tested for simple stringh equivalency. The first of these comparators, the `DateUpdatedComparator`, is added to ensure that comparisons on models that have an audit timetamp incremented every time they are mutated are still able to succeed.

### CODE REVIEW

**Comment by azaslavsky** [comment]
I believe this is ready for review, though for some reason I can't quite figure out it's having trouble recognizing the diff with the previous PR in the stack, so it's lumping it all together. You may have to select just the last commit from the dropdown in the UI to see the proper diff @chadwhitacre

**Comment by chadwhitacre** [comment]
> it's lumping it all together

[Looks like](https://github.com/getsentry/sentry/pull/52484/commits) the commit on #52484 is 26b8d2b5f888feee8419c4d59b84e58cff52fe9a:

Screenshot 2023-07-11 at 9 55 20 AM

... whereas this PR builds on 87e34ac18e751cbcb565210720ac90c0a09807ce:

Screenshot 2023-07-11 at 9 55 13 AM

Perhaps a rebase is in order? Should check #52485 and #52485 as well.

**Comment by azaslavsky** [comment]
The issue was that I had since rebased my local commits, so the hashes had changed. Pushing the other commits in the stack to their respective PRs fixed the issue - this should be ready for review now.

**Comment by chadwhitacre** [comment]
Does d7bf1fc5217e853cbc04a7b74cef233f81488d93 belong on here?

**Comment by chadwhitacre** [comment]
(Maybe that's normal behavior for our workflow, I'm not sure.)

File: src/sentry/utils/env.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import sys
+
+
+def in_test_environment():
+    return sys.argv[0] in {"pytest", "vscode"}
```
Comment (chadwhitacre) [inline src/sentry/utils/env.py:5]: This seems like a fine change but unrelated to the rest of the PR. Ideal would be to put this up as a separate PR for easier/atomic review.

File: tests/sentry/backup/test_correctness.py [hunk 2]
```diff
@@ -2,17 +2,39 @@
...
+        return pytest.mark.django_db(transaction=True, reset_sequences=True, databases="__all__")(
+            func
+        )
+
+    def decorator(function):
+        return pytest.mark.django_db(transaction=True, reset_sequences=True, databases="__all__")(
+            function
+        )
+
+    return decorator
```
Comment (chadwhitacre) [inline tests/sentry/backup/test_correctness.py:34]: Is this in use? Not seeing it.

File: tests/sentry/backup/test_correctness.py [hunk 3]
```diff
@@ -24,24 +46,76 @@ def import_export_then_validate(tmp_path: Path, fixture_file_name: str) -> None:
...
+    import_export_then_validate(tmp_path, "fresh-install.json", DEFAULT_COMPARATORS)
 
 
 @django_db_all(transaction=True, reset_sequences=True)
 def test_bad_fresh_install_validation(tmp_path):
+
     with pytest.raises(ValidationError) as excinfo:
-        import_export_then_validate(tmp_path, "fresh-install.json")
-    assert len(excinfo.value.info.findings) == 2
+        import_export_then_validate(tmp_path, "fresh-install.json", {})
```
Comment (chadwhitacre) [inline tests/sentry/backup/test_correctness.py:67]: Looks goofy to me to pass `{}` as the final arg everywhere. Give the arg a default? But in that case it feels like `DEFAULT_COMPARATORS` should be the default. 🤔

File: tests/sentry/backup/test_correctness.py [hunk 4]
```diff
@@ -24,24 +46,76 @@ def import_export_then_validate(tmp_path: Path, fixture_file_name: str) -> None:
...
+        },
+    }
+    right = {
+        "model": "test",
+        "pk": 1,
+        "fields": {
+            "my_date_field": "2023-06-22T23:00:00.123Z",
+        },
+    }
+    assert cmp.compare(id, left, right) is None
```
Comment (chadwhitacre) [inline tests/sentry/backup/test_correctness.py:98]: Would these be better in a `test_comparators.py` file?

File: src/sentry/runner/commands/backup.py [hunk 5]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
         return "\n".join(f.pretty() for f in self.findings)
 
 
-def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
+class JSONScrubbingComparator(ABC):
+    """An abstract class that compares and then scrubs some set of fields that, by a more nuanced
+    definition that mere strict byte-for-byte equality, are expected to maintain some relation on
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:65]:
```suggestion
    definition than mere strict byte-for-byte equality, are expected to maintain some relation on
```

File: src/sentry/runner/commands/backup.py [hunk 6]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
...
+        if "model" not in data or not isinstance(data["model"], str):
+            raise RuntimeError(f"The {side} input must have a `model` string assigned to it.")
+        if "pk" not in data or not isinstance(data["pk"], int):
+            raise RuntimeError(f"The {side} input must have a numerical `pk` entry.")
+        if "fields" not in data or not isinstance(data["fields"], dict):
+            raise RuntimeError(f"The {side} input must have a `fields` dictionary.")
+
+    @abstractmethod
+    def compare(self, on: InstanceID, left: JSONData, right: JSONData) -> ComparatorFinding | None:
+        """An abstract method signature, to be implemented by inhering classes with their own comparison logic."""
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:92]:
```suggestion
        """An abstract method signature, to be implemented by inheriting classes with their own comparison logic."""
```

File: src/sentry/runner/commands/backup.py [hunk 7]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
...
+
+        pass
+
+    def scrub(self, on: InstanceID, left: JSONData, right: JSONData) -> None:
+        """Removes all of the fields compared by this comparator from the `fields` dict, so that the remaining fields may be compared for equality."""
+
+        self.check("left", left)
+        self.check("right", right)
+        if "scrubbed" not in left:
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:101]: What is `left` here? Do we risk (at least theoretically) that there's already a `scrubbed` key?

File: src/sentry/runner/commands/backup.py [hunk 8]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
...
+            left["scrubbed"] = {}
+        if "scrubbed" not in right:
+            right["scrubbed"] = {}
+        for field in self.fields:
+            del left["fields"][field]
+            left["scrubbed"][f"{self.slug()}::{field}"] = True
+            del right["fields"][field]
+            right["scrubbed"][f"{self.slug()}::{field}"] = True
+
+    def slug(self) -> ComparatorSlug:
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:111]: Would prefer a verb for method name.

File: src/sentry/runner/commands/backup.py [hunk 9]
```diff
@@ -38,7 +39,7 @@ def pretty(self) -> str:
 class ComparatorFinding(NamedTuple):
     """Store all information about a single failed matching between expected and actual output."""
 
-    name: ComparatorName
+    name: ComparatorSlug
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:42]: `slug:`?

File: src/sentry/runner/commands/backup.py (hunk 7 above)
Comment (chadwhitacre) [reply src/sentry/runner/commands/backup.py:101]: I mean, it's a `JSONData`, but that's pretty general. Looking at `check` above, I'm like 80% sure that we're not at risk here (i.e., we're not sharing a namespace with the model fields themselves, i.e., something defined somewhere else that we don't control).

File: src/sentry/runner/commands/backup.py [hunk 10]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
...
+            right["scrubbed"][f"{self.slug()}::{field}"] = True
+
+    def slug(self) -> ComparatorSlug:
+        """A unique identifier for this particular derivation of JSONScrubbingComparator, which will be bubbled up in ComparatorFindings when they are generated."""
+
+        return self.__class__.__name__
+
+
+class DateUpdatedComparator(JSONScrubbingComparator):
+    """Comparator that ensures that the specified field's value on the right input is an ISO-8601 date that is greater than (ie, occurs after) the specified field's left input."""
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/backup.py:118]:
```suggestion
    """Comparator that ensures that the specified field's value on the right input is an ISO-8601 date that is greater than or equal to (ie, occurs on or after) the specified field's left input."""
```

File: src/sentry/utils/env.py (hunk 1 above)
Comment (azaslavsky) [reply src/sentry/utils/env.py:5]: Ah sorry, I did a bad rebase where https://github.com/getsentry/sentry/pull/52775 got merged in!

Comment (azaslavsky) [reply src/sentry/utils/env.py]: Removed in latest push.

File: tests/sentry/backup/test_correctness.py (hunk 2 above)
Comment (azaslavsky) [reply tests/sentry/backup/test_correctness.py:34]: Removed.

File: tests/sentry/backup/test_correctness.py (hunk 4 above)
Comment (azaslavsky) [reply tests/sentry/backup/test_correctness.py:98]: Done.

File: tests/sentry/backup/test_correctness.py (hunk 3 above)
Comment (azaslavsky) [reply tests/sentry/backup/test_correctness.py:67]: I took the default suggestion. I think keeping it `DEFAULT_COMPARATORS` is fine: the `DEFAULT...` refers to production uses, I think it's to be expected that tests might override a production default to validate behavior.

File: src/sentry/runner/commands/backup.py (hunk 5 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:65]: You are incredibly thorough at catching all of the the typos and editing mistakes I make!

File: src/sentry/runner/commands/backup.py (hunk 6 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:92]: Odd that my spell checker missed this one.

File: src/sentry/runner/commands/backup.py (hunk 7 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:101]: `left` and `right` are the two JSON model entries being compared - I've expanded the comment to illuminate this.

In terms of keys already being scrubbed: I think the risk is low, because the actual comparison code near the bottom of the `validate()` function takes care to always run all of the `compare()` methods before any of the `scrub()` methods for exactly this reason, to ensure that each comparator receives data that is unmodified by its predecessor. This means that comparators should be able to be safely listed in `DEFAULT_COMPARATORS` entries in any order without changing the output. I've added a comment to that method below to ensure that future maintainers of this code don't mess up this invariant.

File: src/sentry/runner/commands/backup.py (hunk 9 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:42]: Changed to `kind` and `ComparatorKind`, respectively, for clarity.

File: src/sentry/runner/commands/backup.py [hunk 11]
```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
...
+                reason=f"""the left date_updated value on `{on}` ({left_date_updated}) was not less
+                than or equal to the right ({right_date_updated})""",
+            )
+
+
+ComparatorList = List[JSONScrubbingComparator]
+ComparatorMap = Dict[str, ComparatorList]
+DEFAULT_COMPARATORS: ComparatorMap = {
+    "sentry.userrole": [DateUpdatedComparator("date_updated")],
+    "sentry.userroleuser": [DateUpdatedComparator("date_updated")],
```
[the remark this answers is not in the document]
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:140]: Yes, every model that needs some special comparison logic will need to list its comparators here. I think eventually we could have this logic live on the models themselves, rather than in this file far away from their definitions, but I don't think that sort of clean up is strictly necessary now.

File: src/sentry/runner/commands/backup.py (hunk 8 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py:111]: Changed to `get_kind()`.

File: src/sentry/runner/commands/backup.py (hunk 5 above)
Comment (chadwhitacre) [reply src/sentry/runner/commands/backup.py:65]: > the the typos

Now you're just messing with me! 😅

**Comment by azaslavsky** [comment]
All fixed and cleaned up re: the weird commit merge with the other PR and Mark's PR, and all comments addressed, so should be ready for a second look!

File: src/sentry/runner/commands/backup.py (hunk 5 above)
Comment (azaslavsky) [reply src/sentry/runner/commands/backup.py]: Hahaha I was sure you wouldn't catch it! Pretty impressive how you can spot those so easily - my (and most peoples') eyes just glide right over `the the` and other similar brain-fooling illusions.

Comment (chadwhitacre) [reply src/sentry/runner/commands/backup.py]: I did glide over it the first time! Otherwise I would've mentioned it in my first comment. My tendency is to pore over things, which slows me down but I guess means I catch more typos than average. Tradeoffs! 🙄 🤷‍♂️

File: src/sentry/runner/commands/backup.py (hunk 6 above)
Comment (chadwhitacre) [reply src/sentry/runner/commands/backup.py]: "Inhering" is a word. 🐭

**Review by chadwhitacre — APPROVED** [review]
